nexus/src/app/background/sync_service_zone_nat.rs

```diff
@@ -0,0 +1,233 @@
+        Self { datastore, dpd_clients }
+    }
+}
+
+impl BackgroundTask for ServiceZoneNatTracker {
+    fn activate<'a, 'b, 'c>(
+        &'a mut self,
+        opctx: &'b OpContext,
+    ) -> BoxFuture<'c, serde_json::Value>
+    where
+        'a: 'c,
+        'b: 'c,
+    {
+        async {
+            let log = &opctx.log;
+
+            // check inventory
+            let inventory = match self
+                .datastore
+                .inventory_get_latest_collection(
+                    opctx,
+                    NonZeroU32::new(u32::MAX).unwrap(),
+                )
+                .await
+            {
+                Ok(inventory) => inventory,
+                Err(e) => {
+                    warn!(
+                        &log,
+                        "failed to collect inventory";
+                        "error" => format!("{:#}", e)
+                    );
+                    return json!({
+                        "error":
+                            format!(
+                                "failed collect inventory: \
+                                {:#}",
+                                e
+                            )
+                    });
+                }
+            };
+
+            // generate set of Service Zone NAT entries
+            let collection = match inventory {
+                Some(c) => c,
+                None => {
+                    warn!(
+                        &log,
+                        "inventory collection is None";
+                    );
+                    return json!({
+                        "error": "inventory collection is None"
+                    });
+                }
+            };
+
+            let mut ipv4_nat_values: Vec<Ipv4NatValues> = vec![];
+
+            for (sled_id, zones_found) in collection.omicron_zones {
+                let (_, sled) = match LookupPath::new(opctx, &self.datastore)
+                    .sled_id(sled_id)
+                    .fetch()
+                    .await
+                    .context("failed to look up sled")
+                {
+                    Ok(result) => result,
+                    Err(e) => {
+                        warn!(
+                            &log,
+                            "inventory collection is None";
+                            "error" => ?e,
+                        );
+                        continue;
+                    }
+                };
+
+                let sled_address = external::Ipv6Net(
+                    ipnetwork::Ipv6Network::new(*sled.ip, 128).unwrap(),
+                );
+
+                let zones_config: sled_agent_client::types::OmicronZonesConfig =
+                    zones_found.zones;
+                let zones: Vec<sled_agent_client::types::OmicronZoneConfig> =
+                    zones_config.zones;
+                for zone in zones {
+                    let zone_type: OmicronZoneType = zone.zone_type;
+                    match zone_type {
```

Comment:
ExternalDns is missing from the cases being handled below.

Reply:
Good catch!

Reply:
Fixed

dns zone vnics
---
```
sled03 ➜  omicron git:(rpw-for-service-zone-nat) ✗ pfexec zlogin oxz_external_dns_4fbac8b1-8242-4907-b720-7796c34ecbb3 dladm show-vnic
LINK         OVER         SPEED    MACADDRESS        MACADDRTYPE         VID
vopte4       ?            0        a8:40:25:ff:f7:a9 fixed               0
oxControlService28 ?      0        2:8:20:83:ff:75   random              0
sled03 ➜  omicron git:(rpw-for-service-zone-nat) ✗ pfexec zlogin oxz_external_dns_9ea8ca5a-cb91-40a4-8f10-838424abc1ba dladm show-vnic
LINK         OVER         SPEED    MACADDRESS        MACADDRTYPE         VID
vopte2       ?            0        a8:40:25:ff:c4:3d fixed               0
oxControlService25 ?      0        2:8:20:52:c1:2a   random              0
```

nat entries persisting after restart of dendrite
---
```
root@oxz_switch:~# swadm nat list
External IP  Port low  Port high  Internal IP            Inner MAC          VNI
10.85.0.20   0         65535      fd00:1122:3344:101::1  a8:40:25:ff:c4:3d  100
10.85.0.21   0         65535      fd00:1122:3344:101::1  a8:40:25:ff:f7:a9  100
10.85.0.22   0         65535      fd00:1122:3344:101::1  a8:40:25:ff:9b:eb  100
10.85.0.23   0         65535      fd00:1122:3344:101::1  a8:40:25:ff:86:cd  100
10.85.0.24   0         65535      fd00:1122:3344:101::1  a8:40:25:ff:8f:0b  100
10.85.0.25   0         16383      fd00:1122:3344:101::1  a8:40:25:ff:f6:79  100

root@oxz_switch:~# svcadm restart dendrite
root@oxz_switch:~# swadm nat list
External IP  Port low  Port high  Internal IP            Inner MAC          VNI
10.85.0.20   0         65535      fd00:1122:3344:101::1  a8:40:25:ff:c4:3d  100
10.85.0.21   0         65535      fd00:1122:3344:101::1  a8:40:25:ff:f7:a9  100
10.85.0.22   0         65535      fd00:1122:3344:101::1  a8:40:25:ff:9b:eb  100
10.85.0.23   0         65535      fd00:1122:3344:101::1  a8:40:25:ff:86:cd  100
10.85.0.24   0         65535      fd00:1122:3344:101::1  a8:40:25:ff:8f:0b  100
10.85.0.25   0         16383      fd00:1122:3344:101::1  a8:40:25:ff:f6:79  100
```

Reply:
Thanks. I've retested this in `a4x2` and all the NAT entries came back up. I even rebooted both switches (one at a time) and connectivity came back up on both just fine.